### DISCUSSION TITLE
Project up for Adoption - October 2024

### DISCUSSION THREAD

**Discussion by Jip-Hop** [Jip-Hop/jailmaker#241] [Announcements]
# Jip-Hop Stepping Back as Maintainer of Jailmaker per October 2024

For over 3 years, I have been creating solutions to run `docker` on SCALE. First with a simple [init script](https://www.truenas.com/community/threads/how-to-run-a-docker-container-on-scale-for-dummies.87934/page-3#post-647972). Then when `docker` was removed I explored various alternatives. From [running docker in an LXC container using `libvirt`](https://forum.openmediavault.org/index.php?thread/49272-docker-in-lxc-with-kvm-plugin-apparmor-issue/) to [using `nerdctl`](https://github.com/Jip-Hop/ubernerd) and finally settling on building `jailmaker` around `systemd-nspawn`. [The announcement that Electric Eel will support `docker`](https://forums.truenas.com/t/the-future-of-electric-eel-and-apps/5409) was thus very exciting to me! 🤩 But at the same time it means that I'll personally stop using `jailmaker` and with that I shall also stop maintaining it.

## Thank You!

I'd like to thank the community for your trust in and contributions to `jailmaker` ❤️ and iX for [accepting the request to support `systemd-nspawn`](https://ixsystems.atlassian.net/browse/NAS-119787). We've shown what's possible with `systemd-nspawn` on SCALE and I believe it serves an important role in the Apps transition: [users could migrate to `jailmaker` to continue using docker](https://gist.github.com/Jip-Hop/af3b7a770dd483b07ac093c3b205323f?permalink_comment_id=4428212#gistcomment-4428212) and users who want to continue using `k3s` can start using it.

## Summer Break

Starting this August I shall take a summer break for two months during which I'll be away from GitHub and the TrueNAS forums. Of course you can continue to post in [discussions](https://github.com/Jip-Hop/jailmaker/discussions), create new [issues](https://github.com/Jip-Hop/jailmaker/issues) and [ask questions on the forum](https://forums.truenas.com/t/linux-jails-sandboxes-containers-with-jailmaker/417). There are usually others around to help you out 🙂.

## Future Development

In October I shall try to catch up on everything and update `jailmaker` to be compatible with TrueNAS SCALE Electric Eel 24.10 (in case it doesn't already work in its current form). The last version of `jailmaker` I shall release will probably be [`v3.0.0`, which is already in development](https://github.com/Jip-Hop/jailmaker/tree/v3.0.0). For this release I want to focus on **bugfixes**, **improved testing**, **refactoring** the code and making it **more modular**. I think this will help with the project possibly finding a new maintainer and/or becoming part of the [TrueNAS Middleware](https://github.com/truenas/middleware/) (which is possible as I've released `jailmaker` under the [LGPL-3.0](https://github.com/truenas/middleware/blob/master/LICENSE) open source license as requested by iX).

## Will My Jails Still Work?

This news may come as unwelcome surprise for the people depending on `jailmaker`. While I expect `jailmaker` to continue to function for some time after October 2024, I won't be around to fix it if it doesn't. But rest assured that, even if you'd want to remove the `jailmaker` script from your NAS, [you can still run your current jails just fine without it](https://forums.truenas.com/t/docker-on-truenas-scale/7140/8).

I wish the community of Jailmakers all the best! 🍀

Cheers,
Jip-Hop

**Comment by Lockszmith-GH** [comment]
I must say I've enjoyed this particular ride very much.
I'm thinking, this is still the best way to layer complex infrastructure on top of TrueNAS SCALE when a VM isn't an option.

While you might shift your (future) focus to something else, I would be intrigued to see the direction this project will take after docker will be available on TrueNAS.

Enjoy your Summer Break!

**Reply by Jip-Hop** [reply]
Thanks, it was and is a pleasure working together on this! 🙂

**Comment by jonct** [comment]
You've done a magnificent job, and demonstrated extraordinary vision. Thank you, *thank you*, ***thank you*** for bringing us Jailmaker.

If in any ways big or small I've made this any less fun for you, I sincerely apologize. I hope it's with only poor timing that I bumbled in.

**Reply by Jip-Hop** [reply]
Thanks for the kind words and don't worry about it! I appreciate your contributions.

**Comment by garfunkel** [comment]
I appreciate how you've handled the EOL in a classy and professional way. Great work on this project - you've basically single-handedly freed us all from the treachery and endless breakage of TrueCharts. 🖖

**Comment by aardvarkl** [comment]
@Jip-Hop 
Thank you for what you have done for the last years on jailmaker. It made (at least to me) TrueNAS Scale useable.

**Comment by mrstux** [comment]
@Jip-Hop well done, and enjoy your retirement from Jailmaker maintenance 🎉🥳🎊

EOL though? Surely the project should not actually be declared EOL!

Jailmaker is not just about Docker any more… it’s about Jails and although most people won’t need a jail for docker anymore with EE there are still reasons to make a jail. 

Ie a k3s jail ;)

Or a separate docker instance

or just having some Linux jails

I’m hoping that you transition the jailmaker project to a community project of some sort, not sure how that’s down….  Make a “jailmaker” GitHub organization?

**Reply by sveken** [reply]
Same here, i use jails to separate my docker environments that are used for separate purposes and have different people accessing it.

**Reply by mrstux** [reply]
Yes. I have some docker compose projects which don’t play well with others, and it’s good to have separate jails for each of these.

**Reply by Jip-Hop** [reply]
Well jailmaker by Jip-Hop will be EOL but I hope jailmaker will live on hahah. An organization sounds good but I'm not the one to create it.

**Reply by mrstux** [reply]
Ok. So we’re on the same page! :)

There’s been a fair bit of confusion on Reddit regarding this post. 

Perhaps update the discussion title? 

“JipHop stepping back, looking for maintainers”

;)

**Reply by Jip-Hop** [reply]
Title updated.

**Reply by mrstux** [reply]
As the guy with his face at the top of the readme, I’d adopt the project if wanted at the end of the day

But maybe ix will. Either way. If JipHop is happy to transfer the repo, or whatever the process is, I’m happy to host.

**Reply by jonct** [reply]
I'd follow you. And/or help shoulder any adaptations that iX requests if there's a chance to bring it under their umbrella.

**Reply by Lockszmith-GH** [reply]
Count me in as an active supporter.

**Reply by basnijholt** [reply]
I am also happy to help with the development. I derived a lot of value from the jailmaker script.

After building my new NAS I moved (≈30) almost all of my services from my Proxmox cluster to a jail on TrueNAS and I am really enjoying this new setup.

**Reply by tibmeister** [reply]
Man came across this and really loving it.  Not great at Python currently but willing to learn and also be a steward for this awesome project so others can continue to contribute.  Not sure what @Jip-Hop has in mind for what the adoption would look like and what @mrstux would be able/willing to do.  Will be following closely.

**Reply by mrstux** [reply]
Its October the 15th, and Electric Eel is now imminent. Fangtooth Development is progressing and it appears that iX is integrating LXC/Incus into Fangtooth. There is no sign that nspawn will be removed, and thus Jailmaker should continue to function.

It seems to me that the best course of action is to make it simple for people to migrate their Jailmaker jails to Incus jails, which I think should be fairly simple as the Jailmaker Jail's were actually created using LXC templates.

**Comment by ghan1t** [comment]
Sigh... I haven't even finished migrating my TrueCharts apps to my jailmaker jail.  

What bothers me is that, because this tool is great and easy to use, a lot of people started using and recommending it, reassuring that jails won't go away. Now we might all have to migrate AGAIN.

Of course nobody is entitled to your time on an open source project, but I wish you would have went differently about ending your support for this truly amazing tool, maybe like others suggested searching for other maintainers first? Dropping it like this just after it has started to be used by many, it just kinda sucks. 

That being said, thank you for your time building this, it at least gave people with more complex apps a migration path away from TrueCharts and I appreciate that you maintain it onto Electric Eel.

**Reply by Jip-Hop** [reply]
I would be delighted if one or more people continue the development of jailmaker. And I think publicly announcing that I will stop developing it is the quickest way to get things in motion. But there's no guarantee anyone will volunteer. Therefore I think it's fair to notify the people who are, like yourself, still in the process of migrating ahead of time. Glad you like the tool though and I understand the disappointment.

**Reply by Lockszmith-GH** [reply]
Don't despair @ghan1t - even if the `jlmkr.py` script goes defunct, unlike with the TrueNAS GUI based k3s deployments, nothing is going away, and everything will continue to work.

**Comment by garfunkel** [comment]
You won't have to migrate, or at least it won't be hard.

All this tool does is wrap around the common, standard systemd-nspawn system that comes with Debian/TrueNAS, as well as Docker. What the TrueNAS folk are developing for Electric Eel should allow for either easy migration or basically things to "just work".

This tool - as great as it is - should hopefully be surplus to requirements come Electric Eel.

**Reply by gavdotcom** [reply]
I hope you're right! I truly love jailmaker. On one SCALE box I use it only for docker (my home NAS). On another SCALE box (the work NAS), I use it for a standard systemd-nspawn container of Debian 12 for some commercial backup software (no docker) which works great. 

I've been with this platform since early FreeNAS, and on the BSD variant we had the plugins (like Docker), and jails (like systemd-nspawn), and when SCALE came out and it just had the charts stuff I was like... umm hmmm.... this is fine *burning smell*.  I have friends who won't get off CORE as they are completely tied to jails - they have non-docker apps they want to run natively on their NAS. They'd love to run SCALE but just can't.

Thus it makes complete sense for iXsystems to build in support into the GUI and middleware for both docker (plugins) and the systemd-nspawn (jails). Kinda what I was expecting from the start, and I really hope and pray that Electric Eel is not *just* about docker compose!

**Reply by mrstux** [reply]
I made a video on migrating Docker Sandboxes to the native Docker in Electric Eel, it should explain most things and allay your fears :)

https://www.youtube.com/watch?v=R0Vdj1culo0

**Comment by airbornetrooper82573** [comment]
I didn’t even see this until now and knowing you’re on summer break I just wanted to say thank you. I’ve been able to fully migrate away from Truecharts and I’m in a place ready for electric eel to release. Thank you for creating this project and looking forward to what you come up with next.

**Comment by arpel** [comment]
@Jip-Hop thank you for your very useful (obviously) tool, built in a very professional and user-oriented way (all complexity hidden but accessible if needed). Thanks also for your very quick support to new users ! 
I do believe that the great success and powerfulness of Jailmaker is for something in the "mindset change" of iX regarding Docker support ;)

**Reply by mrstux** [reply]
I think that's a fair call :)

**Comment by rowanmoul** [comment]
@Jip-Hop thank you thank you for all that you have built here. Jailmaker has been a breeze to use from the moment I found it (aside from a few minor bumps involving filesystem permissons and macvlans :wink: ).
I plan to continue using jailmaker for certain apps like Nextcloud-AIO that want direct access to the docker socket, but for everything else I'll probably move to the Truenas built-in docker.

**Comment by Kieros** [comment]
Thank you for your time and effort. It has been fun to work with.
I too will go over to eel with the docker compose.
What will be the best step if you want to retain all the data from previous docks?
Can it be swapped over?

**Reply by Lockszmith-GH** [reply]
This question is more suitable for a discussion of it's own.
But the quick answer is yes. It's all boils down to the volume mappings.

**Reply by mrstux** [reply]
I made a video covering migrating from docker/dockge in a sandbox to native docker on Electric Eel

https://youtu.be/R0Vdj1culo0

**Reply by Lockszmith-GH** [reply]
@mrstux please read #135 and open a new Q&A Discussion with your question.
when you do that, please remove your thread from this discussion - this is not the place for it.

**Comment by clitters** [comment]
I know I am really late to the party ;)
First of all: Thank you for your work you put in. After the desaster with truecharts I think this is or was the best option.

For me personal it still will be, and i hope there will be a maintainer.

One reason why i think this is superior to the electric eel docker-compose approach is, that its way better and easier to implement CI/CD practices with IAC.
I am currently using jailmaker, with a self-hosted Github Runner to deploy my Docker Infrastructure with terraform. Even if its not an option for me to fully deploy two full blown NAS machines, i can spin up my test jail, which has diffrent data mounts, and test my new deployments before merging my code to the main branch witch automatically will deploy it in my "production".

One major issue for me with all docker deployments usually after some time is that its hard to reproduce. As this is not my main job i often only get to work in some spare hours on my projects. i can create a new feature branch, completely mess up my test stage, revert and so on.

Thanks to everyone who supported and improved it to the state it is.
